src/server/run-websocket-server.ts

```diff
@@ -1,29 +1,54 @@
 import * as WebSocket from "ws";
 import * as Knex from "knex";
+import { EventEmitter } from "events";
 import { JsonRpcRequest } from "../types";
 import { isJsonRpcRequest } from "./is-json-rpc-request";
 import { dispatchJsonRpcRequest } from "./dispatch-json-rpc-request";
 import { makeJsonRpcResponse } from "./make-json-rpc-response";
+import { Subscriptions } from "./subscriptions";
 
 export function runWebsocketServer(db: Knex, port: number): void {
   const websocketServer: WebSocket.Server = new WebSocket.Server({ port });
   websocketServer.on("connection", (websocket: WebSocket): void => {
+    const subscriptions = new Subscriptions();
+
     websocket.on("message", (data: WebSocket.Data): void => {
       try {
         const message = JSON.parse(data as string);
-        if (isJsonRpcRequest(message)) {
+        if (!isJsonRpcRequest(message)) return console.error("bad json rpc message received:", message);
+
+        if (message.method === "subscribe") {
+          const eventName: string = message.params.shift();
+
+          try {
+            const subscription: string = subscriptions.subscribe(eventName, message.params, (data: {}): void => {
+              websocket.send(makeJsonRpcResponse(message.id, { subscription, result: data}));
+            });
+            websocket.send(makeJsonRpcResponse(message.id, { subscription }));
+          } catch (exc) {
+            console.error("suscription error", exc, data);
+            websocket.send(makeJsonRpcResponse(message.id, false));
+          }
+        } else if (message.method === "unsubscribe") {
+          const subscription: string = message.params.shift();
+          subscriptions.unsubscribe(subscription);
+          websocket.send(makeJsonRpcResponse(message.id, true));
+        } else {
           dispatchJsonRpcRequest(db, message as JsonRpcRequest, (err?: Error|null, result?: any): void => {
-            if (err) return console.error("dispatch error:", err);
+            if (err) return console.error("dispatch error: ", err);
             websocket.send(makeJsonRpcResponse(message.id, result || null));
           });
-        } else {
-          console.error("bad json rpc message received:", message);
         }
        } catch (exc) {
          console.error("bad json rpc message received:", exc, data);
        }
     });
+
+    websocket.on("close", () => {
+      subscriptions.removeAllListeners();
```

Comment:
Design question: would it be possible/desirable to allow for reconnection attempts before removing all listeners?

Reply:
Possible, yes, desirable -- probably not.

To manage this sort of reconnection logic it requires more bookkeeping on the client and the server.

I think it is in general cleaner to require that we re-bind all listeners after the connection has happened on the client side.

Doing it this way does mean we are way more likely to drop events on the floor in the disconnect/reconnect cycle, and requires that we handle that case if it's an issue.